pin/gc/gc.go

```diff
@@ -33,146 +27,232 @@ type Result struct {
-		err := dag.EnumerateChildren(ctx, getLinks, c, set.Visit)
-		if err != nil {
-			return err
+	defer close(output)
+
+	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
+		links, err := g.ls.GetLinks(ctx, cid)
+		if err != nil && err != dag.ErrNotFound {
+			return nil, &CannotFetchLinksError{cid, err}
 		}
+		return links, nil
 	}
 
-	return nil
-}
-
-// ColoredSet computes the set of nodes in the graph that are pinned by the
-// pins in the given pinner.
-func ColoredSet(ctx context.Context, pn pin.Pinner, ls dag.LinkService, bestEffortRoots []*cid.Cid, output chan<- Result) (*cid.Set, error) {
-	// KeySet currently implemented in memory, in the future, may be bloom filter or
-	// disk backed to conserve memory.
-	errors := false
-	gcs := cid.NewSet()
 	getLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
+		links, err := g.ls.GetLinks(ctx, cid)
 		if err != nil {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+			return nil, &CannotFetchLinksError{cid, err}
 		}
 		return links, nil
 	}
-	err := Descendants(ctx, getLinks, gcs, pn.RecursiveKeys())
+
+	var criticalError error
+	defer func() {
+		if criticalError != nil {
+			output <- Result{Error: criticalError}
+		}
+	}()
+
+	// Enumerate without the lock
+	for {
+		finished, err := tri.EnumerateStep(ctx, bestEffortGetLinks, getLinks)
+		if err != nil {
+			output <- Result{Error: err}
+			criticalError = ErrCannotFetchAllLinks
+		}
+		if finished {
+			break
+		}
+	}
+	if criticalError != nil {
+		return
+	}
+
+	// Add white objects
+	keychan, err := g.bs.AllKeysChan(ctx)
 	if err != nil {
-		errors = true
 		output <- Result{Error: err}
+		return
 	}
 
-	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
-		if err != nil && err != dag.ErrNotFound {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+loop:
+	for {
+		select {
+		case c, ok := <-keychan:
+			if !ok {
+				break loop
+			}
+			tri.InsertFresh(c)
+		case <-ctx.Done():
+			fmt.Printf("ctx done\n")
+			output <- Result{Error: ctx.Err()}
+			return
 		}
-		return links, nil
 	}
-	err = Descendants(ctx, bestEffortGetLinks, gcs, bestEffortRoots)
-	if err != nil {
-		errors = true
-		output <- Result{Error: err}
+
+	// Take the lock
+	unlocker, elock := getGCLock(ctx, g.bs)
+
+	defer unlocker.Unlock()
+	defer elock.Done()
+
+	// Add the roots again, they might have changed
+	for _, r := range g.roots {
+		cids, err := r.Get()
+		if err != nil {
+			criticalError = err
+			return
+		}
+		for _, c := range cids {
+			if !r.Direct {
+				tri.InsertGray(c, r.Strict)
+			} else {
+				// this special case prevents incremental and concurrent GC
+				tri.blacken(c, enumStrict)
+			}
+		}
 	}
 
-	for _, k := range pn.DirectKeys() {
-		gcs.Add(k)
+	// Reenumerate, fast as most will be duplicate
```

Comment:
Not true for `Internal` pins as far as I can tell. Do we use them a lot?

Reply:
No, there will be very little as they are used only for building sharding of the pinset.